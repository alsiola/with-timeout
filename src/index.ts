export type F1<T, U> = (a: T) => Promise<U>;

export const TIMED_OUT = "TIMED_OUT";

function delay(ms: number): Promise<typeof TIMED_OUT> {
    return new Promise(resolve => setTimeout(() => resolve(TIMED_OUT), ms));
}

export class TimeoutError extends Error {
    constructor() {
        super("TIMED_OUT");
    }
}

export function withTimeout<T, U>(fn: F1<T, U>, timeoutMs: number): F1<T, U> {
    return a =>
        new Promise((resolve, reject) => {
            Promise.race([fn(a), delay(timeoutMs)])
                .then(result => {
                    if (result === TIMED_OUT) {
                        return reject(new TimeoutError());
                    }
                    resolve(result);
                })
                .catch(reject);
        });
}
