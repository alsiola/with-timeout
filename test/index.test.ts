import { F1, withTimeout } from "../src";

describe("with-timeout", () => {
    let timeoutFn: F1<string, string>;
    let resolveFn: F1<string, string>;
    let rejectFn: F1<string, string>;

    beforeEach(() => {
        resolveFn = withTimeout((a: string) => new Promise(r => r(a)), 1000);

        rejectFn = withTimeout(
            (a: string) => new Promise((res, rej) => rej(a)),
            1000
        );

        timeoutFn = withTimeout(
            (a: string) => new Promise(r => setTimeout(() => r(a), 200)),
            100
        );
    });

    it("returns fn result if no timeout", () => {
        return expect(resolveFn("test")).resolves.toEqual("test");
    });

    it("throws if timesout", () => {
        return expect(timeoutFn("test")).rejects.toThrowErrorMatchingSnapshot();
    });

    it("passes through fn rejections", () => {
        return expect(rejectFn("fails")).rejects.toBe("fails");
    });

    it("passes arguments for functions with arity > 1", () => {
        const fn = withTimeout(
            (a: string, b: number) => new Promise(r => r([a, b])),
            1000
        );

        expect(fn("test", 5)).resolves.toEqual(["test", 5]);
    });

    it("resolves or time out correctly across multiple uses", () => {
        return Promise.all([
            expect(resolveFn("test")).resolves.toEqual("test"),
            expect(timeoutFn("test")).rejects.toThrowError(),
            expect(timeoutFn("test")).rejects.toThrowError(),
            expect(resolveFn("test")).resolves.toEqual("test"),
            expect(timeoutFn("test")).rejects.toThrowError(),
            expect(timeoutFn("test")).rejects.toThrowError(),
            expect(resolveFn("test")).resolves.toEqual("test"),
            expect(resolveFn("test")).resolves.toEqual("test"),
            expect(timeoutFn("test")).rejects.toThrowError(),
            expect(resolveFn("test")).resolves.toEqual("test"),
            expect(resolveFn("test")).resolves.toEqual("test"),
            expect(timeoutFn("test")).rejects.toThrowError(),
            expect(resolveFn("test")).resolves.toEqual("test"),
            expect(timeoutFn("test")).rejects.toThrowError()
        ]);
    });
});
