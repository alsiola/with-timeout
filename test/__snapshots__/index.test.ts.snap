// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with-timeout throws if timesout 1`] = `"TIMED_OUT"`;
